elks/arch/i86/drivers/char/console-direct.c

```diff
@@ -183,54 +246,114 @@ struct tty_ops dircon_ops = {
     Console_conout
 };
 
-void INITPROC console_init(void)
+
+/* Check to see if this CRTC is present */
+static char probe_crtc(unsigned short crtc_base)
 {
-    Console *C;
     int i;
-    unsigned int PageSizeW;
-
-    MaxCol = (Width = peekb(0x4a, 0x40)) - 1;  /* BIOS data segment */
-
-    /* Trust this. Cga does not support peeking at 0x40:0x84. */
-    MaxRow = (Height = 25) - 1;
-    CCBase = peekw(0x63, 0x40);
-    PageSizeW = ((unsigned int)peekw(0x4C, 0x40) >> 1);
-
-    VideoSeg = 0xb800;
-    if (peekb(0x49, 0x40) == 7) {
-        VideoSeg = 0xB000;
-        NumConsoles = 1;
-        isMDA = 1;
-    } else {
-        isCGA = peekw(0xA8+2, 0x40) == 0;
```

Comment:
As discussed, perhaps leave in the above code when not gated (not using globals though) and call the below code in a separate monitor-ibmpc.c file via a wrapper function. We don't want to change the method of determining the monitor type when DUAL is not configured.

Reply:
That `isCGA` flag is only used in `DisplayCursor` for the CRTC registers `cursor start` and `cursor end`, so it's not really a core part of the display type logic. I'll revisit this when I learn more about EGA and VGA, but I do want to ensure we have a robust setup for detecting this stuff.

Reply:
I was actually referring to the entire bunch of EGA/MDA/CGA init code above: leave that relatively unmodified, then call your external init routine when configured to enhance things if necessary. (I get that some of that isn't possible perhaps given the new design, but that's the general idea).

Specifically about the isCGA flag, this is a good example of somewhat that started simple but got complicated: we wanted to introduce ESC [ ?25 h/l cursor on/off but then found that it didn't work on CGA. Previously, ELKS knew nothing about CGA and now there's another BIOS BDA dependency on it. 

These "BIOS dependencies" are problematic when running on certain hardware emulators (like emu86, etc) and overall contribute to the problems of porting the kernel to embedded systems, PC-98, and the like. Yes, right now we have a completely separate console-direct98.c file, which isn't optimal, but allows @tyama501 total control over getting things to work since he's the ELKS PC-98 features maintainer. At some point the common portions should be merged into a single console-direct.c driver, with only perhaps a specialized init routine for PC-98. In effect, this is what you are kind of doing. I will look further into this thought as you get your code form closer to finalized.

With regards to EGA vs VGA, same issue: ELKS doesn't presently know the difference. We would need to talk more about why we want to know the difference and what programs that would affect before changing kernel behavior between the two. Just more things to think about!